web/modules/custom/collection/src/Plugin/Filter/Glossary.php

```diff
@@ -0,0 +1,248 @@
+use Drupal\filter\Plugin\FilterBase;
+use Drupal\node\NodeInterface;
+use Drupal\og\OgContextInterface;
+use Drupal\og\OgGroupAudienceHelperInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Wraps glossary terms with links to their definition page.
+ *
+ * @Filter(
+ *   id = "collection_glossary",
+ *   title = @Translation("Collection glossary"),
+ *   description = @Translation("Replaces glossary terms with their link version."),
+ *   type = Drupal\filter\Plugin\FilterInterface::TYPE_TRANSFORM_IRREVERSIBLE,
+ * )
+ */
+class Glossary extends FilterBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The OG context service.
+   *
+   * @var \Drupal\og\OgContextInterface
+   */
+  protected $ogContext;
+
+  /**
+   * The entity type manager service.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The collection.
+   *
+   * @var \Drupal\collection\Entity\CollectionInterface
+   */
+  protected $collection;
+
+  /**
+   * The cache metadata to be applied.
+   *
+   * @var \Drupal\Core\Cache\CacheableMetadata
+   */
+  protected $cacheMetadata;
+
+  /**
+   * Constructs a new filter plugin instance.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\og\OgContextInterface $og_context
+   *   The OG context service.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager service.
+   */
+  public function __construct(array $configuration, string $plugin_id, $plugin_definition, OgContextInterface $og_context, EntityTypeManagerInterface $entity_type_manager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->ogContext = $og_context;
+    $this->entityTypeManager = $entity_type_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition): self {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('og.context'),
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function process($text, $langcode): FilterProcessResult {
+    $this->setCollection();
+
+    $result = new FilterProcessResult($text);
+
+    // A collection context cannot be detected.
+    if (!isset($this->collection)) {
+      return $result;
+    }
+
+    // This collection has no glossary term entries.
+    if (!$replacements = $this->getReplacementsMap()) {
+      return $result->addCacheableDependency($this->cacheMetadata);
+    }
+
+    $pattern = '/\b(' . implode('|', array_keys($replacements)) . ')\b/';
```

Comment:
This regex might become a problem when glossaries grow to thousands of terms, but I'm going to let this pass since this is probably never going to happen.